src/Algebra/Graph/Algorithm/Internal.hs

```diff
@@ -1,3 +1,30 @@
 module Algebra.Graph.Algorithm.Internal where
 
+import           Algebra.Graph
+import qualified Data.DisjointSet as DS
+import qualified Data.Set         as S
 
+-- | Extract disjoint set of connectivity components from the graph
+components :: Ord a => Graph a -> DS.DisjointSet a
+components Empty         = DS.empty
+components (Vertex x)    = DS.singleton x
+components (Overlay x y) = mergeComponents (components x) (components y)
+components (Connect x y) = DS.fromLists [vertexList x ++ vertexList y]
+
+-- | Merge two disjoint sets
+mergeComponents ::
+     Ord a => DS.DisjointSet a -> DS.DisjointSet a -> DS.DisjointSet a
+mergeComponents x y = foldr insertComponent x (DS.toSets y)
```

Comment:
`insertComponent` is O(log n), so `mergeComponents` is O(n log n). But I think that it should be possible to make it O(n) if implemented directly on the representation of `DisjointSet`. Let's save it as a suggestion for future improvement for now.